storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala

```diff
@@ -170,8 +195,172 @@ object CharClassPattern extends ParsedPattern {
+              case (char, finalIndex) => Some(CharClassPattern(Set(char)), finalIndex)
+            }
+          }
+          case esc if escapes.contains(esc) =>
+            Some((CharClassPattern(Set(escapes(esc))), startIndex + 1))
+          case other => Some((CharClassPattern(Set(other)), startIndex + 1))
+        }
+      } else {
+        None
+      }
+
+    def addChar(currentChars: List[Char], addChar: Char, inRange: Boolean): List[Char] =
+      if (inRange) {
+        (currentChars.head to addChar).toList ++ currentChars
+      } else {
+        addChar :: currentChars
+      }
+
+    def parseInternalEscapedSequence(startIndex: Int): Option[(Char, Int)] =
+      if (startIndex < str.length) {
+        str.charAt(startIndex) match {
+          case 'x'                          => parseHexChar(startIndex + 1, 2) //escape hex ascii sequence
+          case 'u'                          => parseHexChar(startIndex + 1, 4) //escaped hex unicode sequence
+          case esc if escapes.contains(esc) => Some(escapes(esc), startIndex + 1)
+          case other                        => Some(other, startIndex + 1) //any unsupported escape sequence - just char
+        }
+      } else {
+        None
+      }
+
+    //this function parses interior of a char set [abc-xyz], assuming that
+    //entry '[^' is handled outside, and finished parsing if ']' found
+    //in case of any error, including absent of ']' returns None
+    def parseCharSetSequence(startIndex: Int,
+                             negateCharSet: Boolean): Option[(CharClassPattern, Int)] = {
+      @tailrec
+      def processNextChar(currentChars: List[Char],
+                          currentIndex: Int,
+                          inRange: Boolean): Option[(List[Char], Int)] =
+        if (currentIndex < str.length) {
+          str.charAt(currentIndex) match {
+            case '\\' => {
+              //'\?' escape sequence
+              val charToAdd = parseInternalEscapedSequence(currentIndex + 1)
+              if (charToAdd.isDefined) {
+                //@tailrec restricts flatMap usage here
+                val (nextChar, nextPos) = charToAdd.get
+                processNextChar(addChar(currentChars, nextChar, inRange),
+                                nextPos,
+                                inRange = false)
+              } else {
+                //we got an error during parseInternalEscapedSequence,
+                //for example invalid hex char, or end of string
+                None
+              }
+            }
+            case ']' =>
+              if (inRange) {
+                // closing square bracket when we're in range, something like '[a-]'
+                // that's an error, finish with no result
+                None
+              } else {
+                //we're done
+                Some(currentChars, currentIndex + 1)
+              }
+            case '-' =>
+              if (inRange) {
+                //error found something like "[a--b]"
```

Comment:
The character class [!--b] includes all characters between '!' and '-' together with the character 'b'.

Reply:
Supported with things like []] and other uncommon patterns.